main/src/library/MutDeque.flix

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright 2021 Jakob Schneider Villumsen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use Array.{length, get, put};
+
+/// Represents a mutable deque.
+pub enum MutDeque[a] {
+    case MutDeque(Ref[Array[a]], Ref[Int32], Ref[Int32])
+}
+
+
+namespace MutDeque {
+
+    ///
+    /// Creates a new MutDeque with default values.
+    //
+    pub def new(): MutDeque[a] & Impure = MutDeque(ref [], ref 0, ref 0)
+
+
+    ///
+    /// Removes and returns the front element as an optional of d in O(1) time. Returns None if d is empty.
+    ///
+    pub def popFront(d: MutDeque[a]): Option[a] & Impure = match d {
+        case _ if isEmpty(d) => None
+
+        case MutDeque(arr, f, _) if isWithinBounds(deref f, deref arr) =>
+            let front = deref f;
+            f := increment(front)(withRespectTo(deref arr));
+            Some(get(deref arr, front))
```

Comment:
In addition to incrementing the front index, I think we should set the contents of the backing array at the old front index to `default` (this is a "magic" expression in Flix that evaluates to `null`/`0`/`false` as appropriate). The reason for this is that the garbage collector will still see the old reference to whatever was there and think it's still in use, even though it theoretically doesn't exist anymore.